File: ios/WebyarNative/Sources/Core/Models/ChannelInbox.swift
import SwiftUI

/// A messaging channel the workspace has actually installed — Telegram, Bale,
/// WhatsApp and the rest.
///
/// These are inboxes in the operator's sense but not queues in the server's:
/// the conversations endpoint has no `channel` parameter, and the console
/// narrows its own list the same way. So a channel is a filter laid over
/// whichever queue is open, exactly as `?channel=` is on the web.
struct ChannelInbox: Identifiable, Hashable, Sendable {
    /// The plugin slug, which is also what a conversation carries in
    /// `metadata.channel`.
    let key: String

    var id: String { key }

    /// Written the way the channel writes itself. A product name is not
    /// translated — except Bale, which is Persian to begin with.
    func title(_ language: Language) -> String {
        switch key {
        case "telegram": return "Telegram"
        case "bale": return language == .fa ? "بله" : "Bale"
        case "whatsapp": return "WhatsApp"
        case "instagram": return "Instagram"
        case "x", "twitter": return "X"
        case "messenger", "facebook": return "Messenger"
        case "sms": return language == .fa ? "پیامک" : "SMS"
        // Not plugins, so never in the switcher -- but a conversation can
        // arrive on any of them, and the inbox row names where it came from.
        case "widget": return Str.channelWidget(language)
        case "email": return Str.emailLabel(language)
        case "phone": return Str.channelPhone(language)
        default: return key.prefix(1).uppercased() + key.dropFirst()
        }
    }

    var icon: String {
        switch key {
        case "telegram", "bale", "messenger", "facebook": "paperplane"
        case "whatsapp": "bubble.left.and.bubble.right"
        case "instagram": "camera"
        case "x", "twitter": "at"
        case "sms": "text.bubble"
        case "widget": "bubble.left"
        case "email": "envelope"
        case "phone": "phone"
        default: "square.grid.2x2"
        }
    }

    /// The colour the console gives this channel, so a thread is the same
    /// shade of blue on a phone as on a desk.
    @MainActor
    var tint: Color {
        switch key {
        case "telegram": Color(hue: 200 / 360, saturation: 0.90, brightness: 0.78)
        case "bale": Color(hue: 150 / 360, saturation: 0.60, brightness: 0.62)
        case "whatsapp": Theme.Palette.success
        case "instagram": Color(hue: 330 / 360, saturation: 0.75, brightness: 0.78)
        case "x", "twitter": Theme.Palette.label
        case "widget": Theme.Palette.brand
        default: Theme.Palette.labelSecondary
        }
    }
}

/// One row of `GET /api/plugins/catalog`.
///
/// Only the fields that decide whether a channel belongs in the switcher are
/// decoded; the catalog carries a great deal more that only the marketplace
/// needs.
struct PluginCatalogItem: Decodable, Sendable {
    let slug: String?
    let installed: Bool?
    let supportsInbox: Bool?
    let planAllowed: Bool?
    let installationStatus: String?

    /// Installed, inbox-capable, and still inside the plan.
    ///
    /// `planAllowed` matters as much as `installed`: a workspace that
    /// downgrades keeps its installation row, and an inbox it can no longer
    /// use should not be offered.
    var isUsableInbox: Bool {
        installed == true && supportsInbox == true && planAllowed != false
    }
}

struct PluginCatalogResponse: Decodable, Sendable {
    let items: [PluginCatalogItem]?
}

extension Conversation {
    /// Which channel a thread came in on.
    ///
    /// `metadata.channel`, then `metadata.source`, then the widget — the same
    /// order and the same default as `resolveChannelKey` in the console, so a
    /// conversation is never filed under a different channel on the two
    /// surfaces.
    var channelKey: String {
        if let channel = metadata?["channel"]?.stringValue, !channel.isEmpty { return channel }
        if let source = metadata?["source"]?.stringValue, !source.isEmpty { return source }
        return "widget"
    }
}

File: android/app/src/main/kotlin/com/webyar/operator/core/model/ChannelInbox.kt
package com.webyar.operator.core.model

import com.webyar.operator.i18n.Language
import kotlinx.serialization.Serializable

/**
 * A messaging channel the workspace has actually installed — Telegram, Bale,
 * WhatsApp and the rest.
 *
 * These are inboxes in the operator's sense but not queues in the server's:
 * the conversations endpoint has no `channel` parameter, and the console
 * narrows its own list the same way. So a channel is a filter laid over
 * whichever queue is open, exactly as `?channel=` is on the web.
 */
data class ChannelInbox(
    /**
     * The plugin slug, which is also what a conversation carries in
     * `metadata.channel`.
     */
    val key: String,
) {
    val id: String get() = key

    /**
     * Written the way the channel writes itself.
     *
     * A product name is not translated — except Bale, which is Persian to
     * begin with, and SMS, which is a common noun rather than a brand.
     */
    fun title(language: Language): String = when (key) {
        "telegram" -> "Telegram"
        "bale" -> if (language == Language.FA) "بله" else "Bale"
        "whatsapp" -> "WhatsApp"
        "instagram" -> "Instagram"
        "x", "twitter" -> "X"
        "messenger", "facebook" -> "Messenger"
        "sms" -> if (language == Language.FA) "پیامک" else "SMS"
        else -> key.replaceFirstChar { it.uppercase() }
    }
}

/**
 * One row of `GET /api/plugins/catalog`.
 *
 * Only the fields that decide whether a channel belongs in the switcher are
 * decoded; the catalog carries a great deal more that only the marketplace
 * needs.
 */
/**
 * NOTE THE ABSENCE of `@SerialName` on every field here.
 *
 * `/api/plugins/catalog` answers in camelCase — `supportsInbox`,
 * `planAllowed`, `installationStatus` — while most of this API is snake_case.
 * These were annotated as `supports_inbox` and `plan_allowed` to match the
 * rest, so both decoded to null on every item, `isUsableInbox` was false for
 * every channel, and the switcher was empty on a workspace with Telegram
 * installed and allowed. Nothing failed; the list was simply always empty.
 */
@Serializable
data class PluginCatalogItem(
    val slug: String? = null,
    val installed: Boolean? = null,
    val supportsInbox: Boolean? = null,
    val planAllowed: Boolean? = null,
    val installationStatus: String? = null,
) {
    /**
     * Installed, inbox-capable, and still inside the plan.
     *
     * `planAllowed` matters as much as `installed`: a workspace that
     * downgrades keeps its installation row, and an inbox it can no longer use
     * should not be offered.
     */
    val isUsableInbox: Boolean
        get() = installed == true && supportsInbox == true && planAllowed != false
}

@Serializable
data class PluginCatalogResponse(val items: List<PluginCatalogItem>? = null)

/**
 * Which channel a thread came in on.
 *
 * `metadata.channel`, then `metadata.source`, then the widget — the same order
 * and the same default as `resolveChannelKey` in the console, so a
 * conversation is never filed under a different channel on the two surfaces.
 */
val Conversation.channelKey: String
    get() = metadata.string("channel")?.takeIf { it.isNotEmpty() }
        ?: metadata.string("source")?.takeIf { it.isNotEmpty() }
        ?: "widget"
